### PR TITLE
 feat(pytorch): support Kimi K2.6 

### PR DESCRIPTION
**Pull request by qescccczmr**
## Motivation

Add Kimi-K2.6 support to the PyTorch backend, including text and multimodal inference, compressed-tensors W4A16 MoE, EAGLE3 speculative decoding, and serving integration.

The implementation supports Kimi-K2.6's MLA attention, MoE routing, vision encoder, multimodal projector, and external EAGLE3 draft model.

## Changes

- Add Kimi-K2.6 configuration, model loading, native language model, vision tower, and multimodal projector.
- Add compressed-tensors W4A16 MoE model construction, weight loading, and Triton kernels.
- Add Kimi-K2.6 EAGLE3 speculative decoding with shared target embeddings and auxiliary hidden states.
- Add coverage for compressed-tensors W4A16 kernels, MoE routing, FA3 attention, model loading, and serving.

### Academic

| Dataset | Metric | LMDeploy | HF reference |
| --- | --- | ---: | ---: |
| GPQA Diamond | Accuracy, TP8, 4-run average, CompassAcademic | 91.67% | 90.50% |
| AIME 2026 | Accuracy, TP8 | 97.60% | 96.40% |

LMDeploy and Hugging Face were evaluated with TP8 under aligned evaluation settings.

LMDeploy reaches or slightly exceeds the Hugging Face reference on both academic benchmarks, with no observed text accuracy regression.

### Multimodal

| Dataset | Metric | LMDeploy | HF reference |
| --- | --- | ---: | ---: |
| MathVision | Accuracy | 84.54% | 87.40% |

Kimi-K2.6 successfully completes the MathVision multimodal evaluation.

## Acceptance of EAGLE3

Draft model:

lightseekorg/kimi-k2.6-eagle3.1-mla

### Aggregate results

| Engine | Mean accept length | Strict draft acceptance |
| --- | ---: | ---: |
| LMDeploy | 2.93204 | 64.4015% |
| SGLang | 2.93131 | 64.3769% |
| LMDeploy − SGLang | +0.00074 | +0.0246 pp |

Per-position acceptance rates:

| Engine | Position 0 | Position 1 | Position 2 |
| --- | ---: | ---: | ---: |
| LMDeploy | 81.43% | 63.41% | 48.36% |
| SGLang | 81.52% | 63.21% | 48.40% |

## Runtime Validation

- Kimi-K2.6 TP8 text inference passed on 8 NVIDIA H200 GPUs.
- Kimi-K2.6 TP8+EP8 text inference completed model loading, warmup, and generation successfully.
- EAGLE3 completed all 30 AIME24 requests without engine errors.
- MathVision multimodal evaluation completed successfully.

## Assistance

Assisted with Codex, reviewed manually.

### CODE REVIEW

File: lmdeploy/hf_configs/__init__.py [hunk 1]
```diff
@@ -16,6 +16,9 @@ def register_config(model_type: str):
     elif model_type == 'deepseek_v4':
         from .configuration_deepseek_v4 import DeepseekV4Config
         AutoConfig.register(DeepseekV4Config.model_type, DeepseekV4Config)
+    elif model_type == 'kimi_k2':
```
Comment (grimoire) [inline lmdeploy/hf_configs/__init__.py]: Do we have to register kimi k2 here? https://huggingface.co/moonshotai/Kimi-K2-Thinking/blob/main/config.json the model arch of kimi k2 is deepseek v3.

File: lmdeploy/pytorch/kernels/cuda/fused_noaux_tc.py [hunk 2]
```diff
@@ -95,8 +148,43 @@ def fused_noaux_tc_routing(
     batch_size = logits.shape[0]
     group_size = num_experts // n_group
     assert num_experts % n_group == 0, 'num_experts must be divisible by n_group'
+    use_kimi_kernel = (
```
Comment (grimoire) [inline lmdeploy/pytorch/kernels/cuda/fused_noaux_tc.py]: this is too specialize, could we make the condition more general?

File: lmdeploy/pytorch/kernels/cuda/__init__.py [hunk 3]
```diff
@@ -1,6 +1,7 @@
 # Copyright (c) OpenMMLab. All rights reserved.
 from ..default.w8a8_kernels import per_channel_quant
 from .apply_rotary_pos_emb import apply_rotary_pos_emb
+from .compressed_tensors_w4a16 import fused_moe_w4a16
```
Comment (grimoire) [inline lmdeploy/pytorch/kernels/cuda/__init__.py]: we don't have to expose the kernel here.

File: lmdeploy/pytorch/quantization/__init__.py [hunk 4]
```diff
@@ -0,0 +1,19 @@
+# Copyright (c) OpenMMLab. All rights reserved.
```
Comment (grimoire) [inline lmdeploy/pytorch/quantization/__init__.py]: pytorch/quantization is used to support runtime quantization, I think we might not need to add this feature in this PR.

File: lmdeploy/pytorch/spec_decode/spec_agent.py [hunk 5]
```diff
@@ -186,6 +186,53 @@ def reset_runtime_state(self):
...
+            raise ValueError(
+                f'packed prefill shift expects [1, T] or [1, T, H], got '
+                f'{tuple(input_tensor.shape)}')
+        shifted = input_tensor.clone()
+        if replacement_indices is not None:
+            if replacement_indices.numel() != seq_length.numel():
+                raise ValueError(
+                    'replacement_indices must contain one entry per packed '
+                    'request')
+            replacement_indices = replacement_indices.tolist()
```
Comment (grimoire) [inline lmdeploy/pytorch/spec_decode/spec_agent.py]: tolist would lead to cuda synchronize. Which is not acceptable in our pipeline.

File: tests/pytorch/kernel/test_fa3_attention.py [hunk 6]
```diff
@@ -2,6 +2,7 @@
 import sys
 from types import SimpleNamespace
 
+import pytest
```
Comment (grimoire) [inline tests/pytorch/kernel/test_fa3_attention.py]: We don't need this test, right?

File: lmdeploy/hf_configs/__init__.py (hunk 1 above)
Comment (qescccczmr) [reply lmdeploy/hf_configs/__init__.py]: The target Kimi-K2 model itself does not require this registration because its auto_map points to DeepseekV3Config. However, the standalone EAGLE3 draft model (lightseekorg/kimi-k2.6-eagle3.1-mla) uses model_type: kimi_k2 and does not provide an auto_map, so AutoConfig cannot resolve it when the draft config is loaded independently.

File: lmdeploy/pytorch/configurations/kimi_k2.py [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+            hf_config.aux_hidden_state_layers = (2, cfg.num_layers // 2,
+                                                 cfg.num_layers - 3)
+
+        return cfg
+
+
+# DeepseekV2ModelConfigBuilder also recognizes ``kimi_k2``. Register the
+# narrower Kimi builder immediately before it without changing the shared
+# DeepSeek builder.
+_builders = AutoModelConfigBuilder._sub_classes
```
Comment (grimoire) [inline lmdeploy/pytorch/configurations/kimi_k2.py]: You don't have to patch the builder of dsv2. You can just remove 'kimi_k2' in it's condition.

**Review by grimoire — APPROVED** [review]
LGTM

File: lmdeploy/pytorch/configurations/kimi_k2.py (hunk 7 above)
Comment (qescccczmr) [reply lmdeploy/pytorch/configurations/kimi_k2.py]: I've removed kimi_k2 from DeepseekV2ModelConfigBuilder.

**Comment by grimoire** [comment]
Please fix lint

**Comment by lvhan028** [comment]
what reasoning parser and tool call parser should be for kimi models?

**Comment by qescccczmr** [comment]
> what reasoning parser and tool call parser should be for kimi models?

For Kimi-K2-Thinking, Kimi-K2.5, and Kimi-K2.6 in LMDeploy, use:

```bash
--reasoning-parser default \
--tool-call-parser kimi_k2
```

LMDeploy reuses its unified `default` reasoning parser for Kimi's reasoning delimiters and implicit reasoning-to-tool transition. Kimi native tool calls use a model-specific section containing one or more calls with native IDs and raw JSON arguments, so they remain handled by the dedicated `kimi_k2` tool parser (`kimi-k2` is also accepted as a tool-parser alias).

Moonshot's official [Kimi-K2.6 deployment guide](https://huggingface.co/moonshotai/Kimi-K2.6/blob/7eb5002f6aadc958aed6a9177b7ed26bb94011bb/docs/deploy_guidance.md#vllm-deployment) and [Kimi-K2-Thinking deployment guide](https://huggingface.co/moonshotai/Kimi-K2-Thinking/blob/a51ccc050d73dab088bf7b0e2dd9b30ae85a4e55/docs/deploy_guidance.md#vllm-deployment) use the name `kimi_k2` for the vLLM/SGLang reasoning registry. Parser registry names are engine-specific; the equivalent LMDeploy reasoning implementation is named `default`. The Kimi-K2.6 guide also states that its deployment method can be directly reused from Kimi-K2.5.

Kimi-K2-Instruct is non-reasoning and only needs `--tool-call-parser kimi_k2` when tools are enabled. K2.5/K2.6 Instant mode uses the same server configuration and is selected per request with `{"chat_template_kwargs": {"thinking": false}}`.

File: lmdeploy/serve/parsers/reasoning_parser/kimi_k2_reasoning_parser.py [hunk 8]
```diff
@@ -0,0 +1,19 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+from .reasoning_parser import ReasoningParser, ReasoningParserManager
+
+
+@ReasoningParserManager.register_module(['kimi_k2', 'kimi-k2'])
```
Comment (lvhan028) [inline lmdeploy/serve/parsers/reasoning_parser/kimi_k2_reasoning_parser.py]: Can "default" reasoning parser be reused?

**Comment by lvhan028** [comment]
May also update the supported model list both in README and supported_models.md

File: lmdeploy/serve/parsers/reasoning_parser/kimi_k2_reasoning_parser.py (hunk 8 above)
Comment (qescccczmr) [reply lmdeploy/serve/parsers/reasoning_parser/kimi_k2_reasoning_parser.py]: Yes. I removed the separate Kimi reasoning-parser class and the Kimi reasoning aliases, and now reuse LMDeploy's `default` reasoning parser directly:

```bash
--reasoning-parser default \
--tool-call-parser kimi_k2
```

The unified response parsing already handles Kimi's reasoning delimiters and its implicit transition from reasoning into a tool-call section. K2.5/K2.6 Instant mode remains request-scoped via `{"chat_template_kwargs": {"thinking": false}}`.

Moonshot's [Kimi-K2.6 deployment guide](https://huggingface.co/moonshotai/Kimi-K2.6/blob/7eb5002f6aadc958aed6a9177b7ed26bb94011bb/docs/deploy_guidance.md#vllm-deployment) and [Kimi-K2-Thinking deployment guide](https://huggingface.co/moonshotai/Kimi-K2-Thinking/blob/a51ccc050d73dab088bf7b0e2dd9b30ae85a4e55/docs/deploy_guidance.md#vllm-deployment) use `kimi_k2` as the vLLM/SGLang reasoning-parser name; parser registry names are engine-specific. Updated in 25d1de81.

**Comment by qescccczmr** [comment]
> May also update the supported model list both in README and supported_models.md

Updated the English and Chinese README/model-support lists for Kimi-K2.6 in 3e9d3b74.

**Review by lvhan028 — APPROVED** [review]
